### PR TITLE
Feature/handle cancelled payment

### PR DESCRIPTION
**Pull request by amoriartyCH**
-Added Make Payment controller to handle cancelling of payment at any point
-Added new HTML view for Make Payment controller
-Updated services and existing controllers
-Updated test cases to handle code changes
-Updated Resume link functionality to reflect new journey

[SFA-1739] When i am a user who cancels my payment i want to be taken to a cancel payment screen.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/cic/CICStepsToCompleteController.java [hunk 1]
```diff
@@ -55,7 +55,8 @@ public String postStepsToComplete(@PathVariable String companyNumber,
 
             cicReportService.createCicReport(transactionId, companyAccountsId);
 
-            transactionService.createResumeLink(companyNumber, transactionId, companyAccountsId);
+            String resumeLink = "/company/" + companyNumber + "/transaction/" + transactionId + "/company-accounts/" + companyAccountsId + "/resume";
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/cic/CICStepsToCompleteController.java]: We could construct this from a `UriTemplate`; On the class, define:
```
private static final UriTemplate RESUME_URI =
            new UriTemplate("/company/{companyNumber}/transaction/{transactionId}/company-accounts/{companyAccountsId}/resume");
```

And then to use this, call:
```
RESUME_URI.expand(companyNumber, transactionId, companyAccountsId).toString()
```

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/ApprovalController.java [hunk 2]
```diff
@@ -99,11 +99,15 @@ public String postApproval(@PathVariable String companyNumber,
                 return getTemplateName();
             }
 
-            Optional<String> paymentUrl = transactionService.closeTransaction(transactionId);
-            if (paymentUrl.isPresent()) {
+            boolean paymentRequired = transactionService.closeTransaction(transactionId);
+            if (paymentRequired) {
+
+                String resumeLink = "/company/" + companyNumber + "/transaction/" + transactionId + "/company-accounts/" + companyAccountsId + "/do-you-want-to-make-a-payment";
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/ApprovalController.java]: As above

File: src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/StepsToCompleteController.java [hunk 3]
```diff
@@ -82,7 +82,10 @@ public String postStepsToComplete(@PathVariable String companyNumber,
 
             smallFullService.createSmallFullAccounts(transactionID, companyAccountsID);
             statementsService.createBalanceSheetStatementsResource(transactionID, companyAccountsID);
-            transactionService.createResumeLink(companyNumber, transactionID, companyAccountsID);
+
+            String resumeLink = "/company/" + companyNumber + "/transaction/" + transactionId + "/company-accounts/" + companyAccountsId + "/resume";
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/controller/smallfull/StepsToCompleteController.java]: And here

File: src/main/java/uk/gov/companieshouse/web/accounts/service/payment/PaymentService.java [hunk 4]
```diff
@@ -12,6 +12,6 @@
      * @return A URL to to which to redirect to perform the payment
      * @throws ServiceException if there's an error creating the payment session
      */
-    String createPaymentSessionForTransaction(String transactionId, String paymentUrl)
+    String createPaymentSessionForTransaction(String transactionId)
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/service/payment/PaymentService.java]: The javadoc is no longer relevant for this method

File: src/main/java/uk/gov/companieshouse/web/accounts/service/payment/impl/PaymentServiceImpl.java [hunk 5]
```diff
@@ -49,33 +44,20 @@ public PaymentServiceImpl(ApiClientService apiClientService, EnvironmentReader e
      * {@inheritDoc}
      */
     @Override
-    public String createPaymentSessionForTransaction(String transactionId, String paymentUrl)
+    public String createPaymentSessionForTransaction(String transactionId)
             throws ServiceException {
 
+        String paymentUrl = "/payment";
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/service/payment/impl/PaymentServiceImpl.java]: Can we make this a `private static final String` defined on the class? There's no need to create a new `paymentUrl` every time this method is called.

File: src/main/java/uk/gov/companieshouse/web/accounts/service/transaction/TransactionService.java [hunk 6]
```diff
@@ -24,7 +24,7 @@
      * @return an {@link Optional<String>} containing the URL to be used to create a payment
      *         session, if the transaction is payable
      */
-    Optional<String> closeTransaction(String transactionId) throws ServiceException;
+    boolean closeTransaction(String transactionId) throws ServiceException;
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/service/transaction/TransactionService.java]: Can we update the javadoc inline with this change please?

File: src/main/java/uk/gov/companieshouse/web/accounts/service/transaction/TransactionService.java [hunk 7]
```diff
@@ -35,7 +35,7 @@
      * @param companyAccountsId the company accounts ID of the transaction to update
      * @throws ServiceException
      */
-    void createResumeLink(String companyNumber, String transactionId, String companyAccountsId) throws ServiceException;
+    void updateResumeLink(String transactionId, String resumeLink) throws ServiceException;
```
Comment (BSaunders95) [inline src/main/java/uk/gov/companieshouse/web/accounts/service/transaction/TransactionService.java]: And here

File: src/main/resources/ValidationMessages.properties [hunk 8]
```diff
@@ -16,4 +16,5 @@ consultationWithStakeholdersSelection.selectionNotMade=Tell us if there has been
 directorsRemuneration.directorsRemuneration.missing=Enter the directors'' remuneration information
 directorsRemunerationSelection.selectionNotMade=Tell us if directors'' remuneration was received
 transferOfAssetsSelection.selectionNotMade=Tell us if there there any transfer of assets other than for full consideration
-transferOfAssets.transferOfAssets.missing=Enter details of any transfer of assets that have not reached market value
\ No newline at end of file
+transferOfAssets.transferOfAssets.missing=Enter details of any transfer of assets that have not reached market value
+payment.makePaymentChoice=Please choose whether to make a payment or not
```
Comment (BSaunders95) [inline src/main/resources/ValidationMessages.properties]: You'll need to run the wording choice past content designers. Typically, these statements start 'Tell us...'

File: src/main/resources/templates/payment/makePayment.html [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+<div id="do-you-want-to-make-a-payment-main-content" layout:fragment="content">
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+
+      <form id="do-you-want-to-make-a-payment-form" class="form" th:object="${makePaymentChoice}" th:action="@{''}" method="post">
+
+        <div th:replace="fragments/globalErrors :: globalErrors"></div>
+
+        <h1 id="page-title" class="govuk-heading-l">
+          Do you want to make a payment
```
Comment (BSaunders95) [inline src/main/resources/templates/payment/makePayment.html]: We need a '?' at the end of this question. Same in the page `<title>` tag.

File: src/test/java/uk/gov/companieshouse/web/accounts/controller/payment/MakePaymentControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+import uk.gov.companieshouse.web.accounts.service.payment.PaymentService;
+
+@ExtendWith(MockitoExtension.class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class MakePaymentControllerTest {
+
+    private MockMvc mockMvc;
+
+    @Mock
+    private NavigatorService navigatorService;
```
Comment (BSaunders95) [inline src/test/java/uk/gov/companieshouse/web/accounts/controller/payment/MakePaymentControllerTest.java]: Is this required?

File: src/test/java/uk/gov/companieshouse/web/accounts/controller/payment/MakePaymentControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            .andExpect(view().name(MAKE_PAYMENT_VIEW))
+                .andExpect(model().attributeExists(MAKE_PAYMENT_MODEL_ATTR));
+    }
+
+    @Test
+    @DisplayName("Make payment post request - Yes")
+    void postMakePaymentYesChoice() throws Exception {
+        mockMvc.perform(post(MAKE_PAYMENT_PATH).
+            param(MAKE_PAYMENT_MODEL_ATTR, "1"))
+            .andExpect(status().is3xxRedirection());
```
Comment (BSaunders95) [inline src/test/java/uk/gov/companieshouse/web/accounts/controller/payment/MakePaymentControllerTest.java]: We could assert `view().name` here

File: src/test/java/uk/gov/companieshouse/web/accounts/service/transaction/impl/TransactionServiceImplTests.java [hunk 12]
```diff
@@ -1,5 +1,5 @@
 package uk.gov.companieshouse.web.accounts.service.transaction.impl;
-
+/*
```
Comment (BSaunders95) [inline src/test/java/uk/gov/companieshouse/web/accounts/service/transaction/impl/TransactionServiceImplTests.java]: These tests need to be uncommented and updated inline with new functionality

Comment (amoriartyCH) [reply src/test/java/uk/gov/companieshouse/web/accounts/service/transaction/impl/TransactionServiceImplTests.java]: Implemented all suggested changes. Will still need to confirm with someone about the wording of the validation message (but have updated it to begin with the words "Tell us").

File: src/main/resources/templates/payment/payFilingFee.html [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    <div class="govuk-grid-column-two-thirds">
+
+      <form id="do-you-want-to-make-a-payment-form" class="form" th:object="${payFilingFeeChoice}" th:action="@{''}" method="post">
+
+        <div th:replace="fragments/globalErrors :: globalErrors"></div>
+
+        <h1 id="page-title" class="govuk-heading-l">
+          Do you want to pay the filing fee now?
+        </h1>
+        <p>Payment for this filing was not completed. Your filing has not been submitted.</p>
```
Comment (BSaunders95) [inline src/main/resources/templates/payment/payFilingFee.html]: This will need an id

**Pull request merged**